driver/generic/st77xx.py

```diff
@@ -0,0 +1,415 @@
+    const(ST77XX_MADCTL_MY | ST77XX_MADCTL_MV), # 3 = inverted landscape
+)
+
+ST77XX_COLOR_MODE_65K = const(0x50)
+ST77XX_COLOR_MODE_262K = const(0x60)
+ST77XX_COLOR_MODE_12BIT = const(0x03)
+ST77XX_COLOR_MODE_16BIT = const(0x05)
+ST77XX_COLOR_MODE_18BIT = const(0x06)
+ST77XX_COLOR_MODE_16M = const(0x07)
+
+
+ST77XX_COL_ROW_MODEL_START_ROTMAP={
+    # ST7789
+    (240,320,None):[(0,0),(0,0),(0,0),(0,0)],
+    (240,240,None):[(0,0),(0,0),(0,80),(80,0)],
+    (135,240,None):[(52,40),(40,53),(53,40),(40,52)],
+    # ST7735
+    (128,160,'blacktab'):[(0,0),(0,0),(0,0),(0,0)],
+    (128,160,'redtab'):[(2,1),(1,2),(2,1),(1,2)],
+}
+
+ST77XX_PORTRAIT = const(0)
+ST77XX_LANDSCAPE = const(1)
+ST77XX_INV_PORTRAIT = const(2)
+ST77XX_INV_LANDSCAPE = const(3)
+
+class St77xx_hw(object):
+    def __init__(self, *, cs, dc, spi, res, suppRes, bl=None, model=None, suppModel=[], rst=None, rot=ST77XX_LANDSCAPE, bgr=False, rp2_dma=None):
+        '''
+        This is an abstract low-level driver the ST77xx controllers, not to be instantiated directly.
+        Derived classes implement chip-specific bits. THe following parameters are recognized:
+
+        * *cs*: chip select pin (= slave select, SS)
+        * *dc*: data/command pin
+        * *bl*: backlight PWM pin (optional)
+        * *model*: display model, to account for variations in products
+        * *rst*: optional reset pin
+        * *res*: resolution tuple; (width,height) with zero rotation
+        * *rot*: display orientation (0: portrait, 1: landscape, 2: inverted protrait, 3: inverted landscape); the constants ST77XX_PORTRAIT, ST77XX_LANDSCAPE, ST77XX_INV_POTRAIT, ST77XX_INV_LANDSCAPE may be used.
+        * *bgr*: color order if BGR (not RGB)
+        * *rp2_dma*: optional DMA object for the rp2 port
+
+
+        Subclass constructors (implementing concrete chip) set in addition the following, not to be used directly:
+
+        * *suppModel*: models supported by the hardware driver
+        * *suppRes*: resolutions supported by the hardware driver, as list of (width,height) tuples
+        '''
+        self.buf1 = bytearray(1)
+        self.buf2 = bytearray(2)
+        self.buf4 = bytearray(4)
+
+        self.cs,self.dc,self.bl,self.rst=[(machine.Pin(p,machine.Pin.OUT) if isinstance(p,int) else p) for p in (cs,dc,bl,rst)]
+        self.bl.value(1)
+        self.bl=machine.PWM(self.bl)
+        self.rot=rot
+        self.bgr=bgr
+        self.width,self.height=(0,0) # set in apply_rotation
+
+        if res not in suppRes: raise ValueError('Unsupported resolution %s; the driver supports: %s.'%(str(res),', '.join(str(r) for r in suppRes)))
+        if suppModel and model not in suppModel: raise ValueError('Unsupported model %s; the driver supports: %s.'%(str(model),', '.join(str(r) for r in suppModel)))
+
+        self.res=res
+        self.model=model
+
+        self.rp2_dma=rp2_dma
+        self.spi=spi
+        self.hard_reset()
+
+    def off(self):
+        self.bl.value(0)
```

Comment:
Even if `bl` is not `None` - this would still crash because you've set `self.bl` to `machine.PWM` which doesn't have "value" function.  
If you want PWM at least name it differently instead of re-assigning `self.bl` to a different type. This would help prevent confusion.

Reply:
Same as above, thank's for catching that. I will test with `bl=None` later.